lib/unlock-helpers.js

```diff
@@ -0,0 +1,185 @@
+  }
+  if (type === PIN_UNLOCK) {
+    return /^[0-9]+$/.test(key.trim());
+  }
+  if (type === PATTERN_UNLOCK) {
+    if (!/^[0-9]+$/.test(key.trim())) {
+      return false;
+    }
+    return !(/([0-9]).*?\1/.test(key.trim()));
+  }
+  // Dont trim password key, you can use blank spaces in your android password
+  // ¯\_(ツ)_/¯
+  if (type === PASSWORD_UNLOCK) {
+    return /.{4,}/g.test(key);
+  }
+  throw new Error(`Invalid unlock type ${type}`);
+};
+
+helpers.dismissKeyguard = async function (driver, adb) {
+  let isKeyboardShown = await driver.isKeyboardShown();
+  if (isKeyboardShown) {
+    await driver.hideKeyboard();
+    await sleep(100);
+  }
+  // dismiss notifications
+  logger.info("Dismiss notifications from unlock view");
+  await adb.shell(["service", "call", "notification", "1"]);
+  await adb.back();
+  if (await adb.getApiLevel() > 21) {
+    logger.info("Trying to dismiss keyguard");
+    await adb.shell(["wm", "dismiss-keyguard"]);
+    return;
+  }
+  logger.info("Swiping up to dismiss keyguard");
+  await helpers.swipeUp(driver);
+};
+
+helpers.swipeUp = async function (driver) {
+  let windowSize = await driver.getWindowSize();
+  let x0 = parseInt(windowSize.x / 2);
+  let y0 = windowSize.y - 10;
+  let yP = 100;
+  let actions = [
+    {action: 'press', options: {element: null, x: x0, y: y0}},
+    {action: 'moveTo', options: {element: null, x: x0, y: yP}},
+    {action: 'release'}
+  ];
+  await driver.performTouch(actions);
+};
+
+helpers.encodePassword = function (key) {
+  return key.replace(/\s+/ig, "%s");
+};
+
+helpers.pinUnlock = async function (adb, driver, capabilities) {
+  logger.info(`Trying to unlock device using pin ${capabilities.unlockKey}`);
+  await helpers.dismissKeyguard(driver, adb);
+  let key = capabilities.unlockKey.split("");
+  if (await adb.getApiLevel() >= 21) {
+    let els = await driver.findElOrEls("id", "com.android.systemui:id/digit_text", true);
+    if (els === null || els.length === 0) {
+      throw new Error("Error finding unlock pin buttons!");
+    }
+    let pins = {};
+    for (let e in els) {
+      let text = await driver.getAttribute("text", els[e].ELEMENT);
+      pins[text] = els[e];
+    }
+    for (let i in key) {
+      let pin = key[i];
+      let el = pins[pin];
+      await driver.click(el.ELEMENT);
+    }
+    let el = await driver.findElOrEls("id", "com.android.systemui:id/key_enter", false);
+    await driver.click(el.ELEMENT);
+  } else {
+    for (let i in key) {
+      let pin = key[i];
+      let el = await driver.findElOrEls("id", `com.android.keyguard:id/key${pin}`, false);
+      if (el === null) {
+        throw new Error("Error finding unlock pin button!");
+      }
+      await driver.click(el.ELEMENT);
+    }
+    let el = await driver.findElOrEls("id", "com.android.keyguard:id/key_enter", false);
+    await driver.click(el.ELEMENT);
+  }
+  await sleep(100);
+};
+
+helpers.passwordUnlock = async function (adb, driver, capabilities) {
```

Comment:
no real reason to pass all the caps in here, right? when all we need is `unlockKey`?